File: 2015/solutions/javascript/day4.js
const crypto = require("crypto")

module.exports = (class Solution extends require("./solution.js") {
    findIndexOfHashMatching(predicate) {
        for (let i = 0;; i++)
            if (predicate(crypto.createHash("md5").update(this.input + i).digest("hex")))
                return i;
    }

    getSeed(startsWith) {
        return this.findIndexOfHashMatching(hash => hash.startsWith(startsWith))
    }

    partOne() {
        return this.getSeed("00000");
    }

    partTwo() {
        return this.getSeed("000000");
    }
}).instantiate(4);

File: 2015/solutions/python/day_4.py
#!/usr/bin/env python3

from itertools import count
from hashlib import md5

from solution import SolutionBase

def find_hash(secret, match_function):
    inputs = (f"{secret}{x}".encode("ascii") for x in count())
    hashed = (md5(hash_input).hexdigest() for hash_input in inputs)
    pairs = zip(count(), hashed)
    valid = filter(lambda x: match_function(x[1]), pairs)
    return next(valid)[0]

class Solution(SolutionBase):
    def part_one(self):
        matches = lambda hashed: str(hashed).startswith("00000")
        return find_hash(self.input, matches)

    def part_two(self):
        matches = lambda hashed: str(hashed).startswith("000000")
        return find_hash(self.input, matches)

solution = Solution.instantiate(4)

if __name__ == "__main__":
    solution.run()
